New Sched attribute to throttle job attribute updates + refactoring

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
Scheduler sends job attribute updates for certain attributes to the server. These updates can become a performance bottleneck, affecting scheduling speed. The most problematic ones are the updates related to “job can’t run”, where scheduler updates the job comment, possibly accrue_type, for each job that’s not run, even if it might run those jobs in the very next cycle. So, throttling such updates  sounds reasonable, and can speed up the scheduler greatly. By reducing the total number of requests sent to the server, it should also help the server be more responsive to other more important requests.

So, adding a sched attribute to do the same. Please see design document for more details.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
- Added new sched attribute "attr_update_period" to control after how long since last the scheduler will send attribute updates to server for "can't run" type updates.
- Refactored the code so that scheduler attributes are not queried every scheduling cycle, they are all cached by sched and updated when one updates them via qmgr.
- Removed sched_config's hidden option 'opt_backfill_fuzzy_time' and enhanced sched attribute opt_backfill_fuzzy to accept time in seconds in addition to existing values "low", "med" etc.

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->
https://pbspro.atlassian.net/wiki/spaces/PBSPro/pages/1684045848/New+Sched+attribute+to+throttle+job+attribute+updates

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
[sched_attr_updates.log](https://github.com/PBSPro/pbspro/files/4526871/sched_attr_updates.log)

(Performance test results will be attached soon)


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

bhroam commented on src/scheduler/fifo.c:
```diff
@@ -657,6 +654,12 @@ scheduling_cycle(int sd, char *jobid)
 
+	/* Decide whether we need to send "can't run" type updates this cycle */
+	if (time(NULL) - last_attr_updates > sc_attrs.attr_update_period)
+		send_cant_run = 1;
```
Rename this to send_job_attr_updates

agrawalravi90 commented on src/scheduler/constant.h:
```diff
@@ -614,6 +614,12 @@ enum sort_info_type {
+	PS_MIN_T_SINCE_START,
+	PS_PREEMPT_PRIORITY,
+	PS_HIGH
+};
```
@tpai26 are you going to remove it anytime soon?

tpai26 replied:
@agrawalravi90 its going to be removed in a couple of weeks

tpai26 replied:
hi, just an update on this-  this attribute is not going to be removed anymore after discussions, there migt be requirement in future to add more alues to this attribute - https://pbspro.atlassian.net/wiki/spaces/PD/pages/1747058710/preempt+sort+attribute+will+only+have+one+value

agrawalravi90 replied:
great, good thing I never got rid of it in my branch then, and this enum will help if we add more values, thanks for the update!

agrawalravi90 commented on src/scheduler/fifo.c:
```diff
@@ -469,13 +469,13 @@ init_scheduling_cycle(status *policy, int pbs_sd, server_info *sinfo)
 						   float_digits(resresv->job->formula_value, FLOAT_NUM_DIGITS), resresv->job->formula_value);
 
-					if (!resresv->can_not_run && policy->job_form_threshold_set && resresv->job->formula_value <= threshold) {
+					if (!resresv->can_not_run && resresv->job->formula_value <= threshold) {
```
I didn't realize, thanks for catching that

agrawalravi90 commented on src/server/pbsd_init.c:
```diff
@@ -616,7 +616,8 @@ pbsd_init(int type)
 			dflt_scheduler = sched_alloc(PBS_DFLT_SCHED_NAME);
 			set_sched_default(dflt_scheduler, 0);
-			(void)sched_save_db(dflt_scheduler, SVR_SAVE_NEW);
+			set_scheduler_flag(SCH_ATTRS_CONFIGURE, dflt_scheduler);
```
I thought it would be safe to do this, just in case there were any race conditions before server went down the last time. Do you strongly feel that we should remove it?

bhroam replied:
That's a pretty small race crash condition.  The server would have to crash between when the sched was saved to the DB and the next time through the main loop when the high priority scheduling cycle happened.  If you want this to happen, call the update attrs call in SCH_SCHEDULE_FIRST case in schedule().  This happens when the server first comes up.

agrawalravi90 replied:
Ok, ya that might be better, thanks.

bhroam commented:
I was thinking about race conditions and there is one in the server.  There are multiple high priority scheduling commands, but only one variable in the server.  This means one can overwrite the other.  To help solve this, maybe combine SCH_CONFIGURE_ATTRS and SCH_CONFIGURE.  I know I told you to not update the attributes in a SCH_CONFIGURE, but I think you can check if the pbs_sd is valid before you call it.  It does mean that if you set something via qmgr, the scheduler will reread the sched_config, but I doubt that is a big deal.  It'll save us from a race in the server.

agrawalravi90 commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -0,0 +1,140 @@
+        self.server.manager(MGR_CMD_SET, SCHED, a, id="default")
+
+        j = Job()
+        j.set_sleep_time(1000)
```
I need the first job to remain in the system and running until the test is over, that way all the other jobs will accrue eligible time. I was afraid that 100 seconds (default) might not be enough on some slow machines, so set it to 1000

agrawalravi90 commented:
@bhroam I've addressed your comments, please provide further feedback, thanks!

bhroam commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -132,9 +130,11 @@ def test_accrue_type(self):
-        self.server.expect(JOB, "accrue_type", op=SET, id=jid2)
-        self.server.expect(JOB, {"eliglible_time": "00:00:00"}, op=NE, id=jid2)
+        ret = self.server.status(JOB, id=jid4)
+        elig_time = ret[0]["eligible_time"]
```
Why the change from expect() to status()?

agrawalravi90 replied:
thanks for pointing it out. The expect was failing and I couldn't figure out why because eligible_time was correctly getting accrued. Now, finally, I see that I had a typo "eliglible" instead of "eligible", doh! I'll change it back.

agrawalravi90 commented on src/scheduler/fifo.c:
```diff
@@ -2487,14 +2484,11 @@ parse_sched_obj(struct batch_status *status)
-		init_sc_attrs();
-		firsttime = 0;
-	}
+	init_sc_attrs();
```
Not sure I understand, the attrs are only updated on qmgr command runs right (not in main sched loop)? We need to do this every time because we don't set "default" value for all sched attributes when we unset them. So, when something is unset, we might not update the respective attribute unless we call this function.

bhroam replied:
I'm not talking about main_sched_loop().  I'm talking about the loop in main() in pbs_sched.c.  The first thing it does is make a call to update the scheduler attrs.  This is meant to happen only on the first cycle after we start.  This is what firstcycle is meant for.  It's so that call just does something on the first cycle.

agrawalravi90 replied:
ah, ok, thanks for clarifying. But I think we still need to call init_sc_attrs() every time a qmgr operation on sched object is done. init_sc_attrs() just resets all sc_attrs to their default, so that if any sched attribute has been unset, sched will set it to its default. Take job_sort_formula_threshold for example, it doesn't have a default. So, if admin unsets it, the batch_status list will not have a default value of job_sort_formula_threshold, so if we don't call init_sc_attrs() then we will leave it at its old value. Does that make sense?

agrawalravi90 commented on src/scheduler/fifo.c:
```diff
@@ -2507,13 +2501,14 @@ parse_sched_obj(struct batch_status *status)
-				sc_attrs.attr_update_period = newval;
+			sc_attrs.attr_update_period = newval;
+			if (newval != prev_attr_u_period) {
 				last_attr_updates = 0;
```
i think it might be less confusing if we reset it, e.g - if it's been 4 mins since last update, admin sets the period to 5 mins but we send the first update out after 1 min, it might not be obvious why. Secondary reason: It makes it easier to write automated tests. Let me know if you still feel that we should not do this.

bhroam replied:
I don't agree that it is less confusing.  How I see it is that you are punishing jobs that have been there longer.  If you constantly change this period one second before it triggers, you will never update any jobs.  If you keep it the same, you'll either immediately trigger because you lowered the period, or wait a little longer if you raised it.  You are correct, it is harder to write automated tests.  You can force it to be reset, but it is kind of hacky.  You can set it to 0 and run a cycle, and then set it to what you want.

agrawalravi90 replied:
" How I see it is that you are punishing jobs that have been there longer. If you constantly change this period one second before it triggers, you will never update any jobs"

No, we will send out updates when last_attr_updates is 0, so if somebody constantly changes the period one second before it triggers, they will have just triggered it manually. The idea is: if you modify the period, it will be like it's been set for the first time: we will send out the first update, then we will send it N seconds later.

bhroam replied:
Ahh OK, I misunderstood.  I thought we waited the period seconds before sending the updates.  I'm fine with this now.

agrawalravi90 commented:
"I still see the log message. Were you planning on deleting it here?"

I've found it useful for debugging purposes. I think we should keep it, maybe at level DEBUG2 so that it doesn't pollute the logs by default. What do you think?

bhroam commented:
I disagree.  This is just one attribute among many.  I think you only find it useful because it is the current attribute you are developing.  It is already logged in the server log, so it doesn't need to be logged here.

agrawalravi90 commented:
@bhroam gentle reminder, did you get a chance to look at the latest commit?

bhroam commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -0,0 +1,140 @@
+# trademark licensing policies.
+
+from tests.functional import *
+import pprint
```
I didn't see this used.

agrawalravi90 commented on src/include/pbs_internal.h:
```diff
@@ -190,6 +190,9 @@ extern "C" {
 #define PBS_PREEMPT_SORT_DEFAULT	"min_time_since_start"
 
+/* Default value of sched attribute "attr_update_freq" */
+#define SCHED_ATTR_UPDATE_PERIOD_DFLT	"0"
```
I think it's better to be consistent and have defaults which show up in stats even if the attribute is unset. Do let me know if you feel strongly against it.

bhroam replied:
I personally think it adds to clutter.  This will be an attribute that most sites will ignore.  There is no reason to have it show up for them.

agrawalravi90 commented on src/scheduler/fifo.c:
```diff
@@ -2450,39 +2478,174 @@ sched_settings_frm_svr(struct batch_status *status)
+			else if (!strcasecmp(attrp->value, "high"))
+				sc_attrs.opt_backfill_fuzzy = BF_HIGH;
+			else
+				sc_attrs.opt_backfill_fuzzy = BF_DEFAULT;
```
Thanks for mentioning that! it wasn't working with the config option

agrawalravi90 commented on src/scheduler/fifo.c:
```diff
@@ -2692,7 +2855,7 @@ update_svr_schedobj(int connector, int cmd, int alarm_time)
 	if (cmd == SCH_ERROR || connector < 0)
 		return 1;
 
```
As we discussed offline, parse_sched_obj should only get called once when it first receives a command from the server. So, I think we should be good here, but please let me know if you still think that I need to make any changes here.

agrawalravi90 commented on src/scheduler/fifo.c:
```diff
@@ -2753,7 +2916,7 @@ update_svr_schedobj(int connector, int cmd, int alarm_time)
  */
 int
-validate_sched_attrs(int connector)
+set_validate_sched_attrs(int connector)
```
It isn't called on a HUP, but it's a good thing to check so I'll add it, thanks.

bhroam replied:
schedule() with SCH_CONFIGURE is called on a HUP.  That calls set_validate_sched_attrs().

agrawalravi90 commented on src/scheduler/job_info.c:
```diff
@@ -1695,6 +1722,9 @@ update_job_attr(int pbs_sd, resource_resv *resresv, char *attr_name,
 			return -1;
 	} else {
+		if ((flags & UPDATE_NOW) && !can_send_update(attr_name))
+			return 0;
```
Oh, good catch, I didn't realized this

agrawalravi90 commented on src/server/sched_func.c:
```diff
@@ -964,39 +941,42 @@ set_sched_default(pbs_sched *psched, int from_scheduler)
 
-
-	if (flag)
-		set_scheduler_flag(SCH_ATTRS_CONFIGURE, psched);
```
Thanks for mentioning this, I did miss one place. Since it needs to happen every time set_sched_default is called, I think it makes more sense to move it back to set_sched_default().

agrawalravi90 commented on src/tools/pbs_python.c:
```diff
@@ -320,11 +320,6 @@ action_sched_log(attribute *pattr, void *pobj, int actmode)
-action_sched_log_events(attribute *pattr, void *pobj, int actmode)
-{
-	return 0;
-}
```
I did remove it right? did you mean a different function?

bhroam replied:
I did indeed.  Sorry about that.  I meant action_sched_server_dyn_res_alarm() since you removed both of these functions.  Was it not in pbs_python?

agrawalravi90 replied:
I don't see it .. even I was surprised with this, I thought this would cause compilation issues

agrawalravi90 commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -0,0 +1,140 @@
+                              starttime=t)
+
+        # Sleep for 60 seconds
+        time.sleep(60)
```
I decided to use 60 as my attr_update_frequency to make sure that it is enough for slower machines. I'll add the logger info

bhroam replied:
I still think 60 is too long even for slower machines.  Probably 30 is fine.  If you disagree, maybe we can compromise at 45?

agrawalravi90 commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -0,0 +1,140 @@
+        self.server.expect(JOB, {"eligible_time": "00:00:00"}, op=NE, id=jid3)
+        self.server.expect(JOB, "accrue_type", op=SET, id=jid2,
+                           max_attempts=5)
+        self.server.expect(JOB, {"eligible_time": "00:00:00"}, op=NE, id=jid2)
```
Thanks for pointing this out

bhroam commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -81,9 +81,9 @@ def test_basic_throttling(self):
-        self.logger.info("Going to sleep for 60 seconds")
-        time.sleep(60)
+        # Sleep for 45 seconds
+        self.logger.info("Going to sleep for 45 seconds")
```
Remove the comment, it is not needed with the log message.  Also, I really do want you to explain why you are sleeping.

agrawalravi90 commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -119,22 +121,19 @@ def test_accrue_type(self):
-        self.server.expect(JOB, "comment", op=SET, id=jid4)
-        self.server.expect(JOB, "accrue_type", op=SET, id=jid4)
-        self.server.expect(JOB, {"eligible_time": "00:00:00"}, op=NE, id=jid4)
+        self.server.expect(JOB, "comment", op=SET, id=jid4, max_attempts=5)
```
TBH, I think max_attempts should be set to 1, we want to make sure that it was set in the scheduling cycle that just happened. If it gets set in future cycles, it doesn't help us verify whether attribute throttling is working correctly or not.

agrawalravi90 commented on src/scheduler/data_types.h:
```diff
@@ -1150,6 +1150,7 @@ struct config
 	unsigned resv_conf_ignore:1;  /* if we want to ignore dedicated time when confirming reservations.  Move to enum if ever expanded */
 	unsigned allow_aoe_calendar:1;        /* allow jobs requesting aoe in calendar*/
 	unsigned logstderr:1;               /* log to stderr as well as log file */
+	unsigned opt_bf_fuzzy_time_set:1;	/* sched config opt_backfill_fuzzy_time is set */
```
That's what I had done before, and it created a catch-22 situation. We need to know inside init_sc_attrs whether sched_config's opt_backfill_fuzzy_time has been set or not, otherwise we won't be able to set it to the default (BF_DEFAULT) correctly. Let me know what you think.

I'll remove logstderr thanks

bhroam replied:
I'm a little confused.  If opt_backfill_fuzzy_time is set to BF_DEFAULT, then why can't it be used as the default from that point forward?

agrawalravi90 replied:
Consider the case when pbs_sched comes up, we first call schedinit(), which, if opt_backfill_fuzzy_time has been set in sched_config, will set sc_attrs.opt_backfill_fuzzy to that time. Then, when sched receives its first command from server, it calls init_sc_attrs(), if we set sc_attrs.opt_backfill_fuzzy to BF_DEFAULT there then we'll overwrite the sched_config value.

bhroam replied:
I don't think I was clear.  You can't get rid of conf.opt_backfill_fuzzy_time.  Set this variable to BF_DEFAULT in init_config().  If opt_backfill_fuzzy_time is in the config file, set it to that.  In init_scattrs(), set scattrs.opt_backfill_fuzzy_time to conf.opt_backfill_fuzzy_time.  If opt_backfill_fuzzy is not set by in the sched object, we'll use that.

agrawalravi90 replied:
Ah, ok, I see your point. I was going to do that earlier, but I thought that it would be more clear if we set sc_attrs.opt_backfill_fuzzy directly in init_config() instead of having 2 different globals for it, which are set by different code paths. So, I've removed conf.opt_backfill_fuzzy_time and now only have a flag which says whether sc_attrs.opt_backfill_fuzzy was set by init_config() or not. I thought that'd be more clear, let me know what you think.

agrawalravi90 replied:
parse_config* not init_config, sorry for the confusion

bhroam commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -81,8 +81,9 @@ def test_basic_throttling(self):
-        self.logger.info("Going to sleep for 45 seconds")
+        self.logger.info(
+            "Going to sleep for 45 seconds to allow attr_update_period "
+            "to be lapsed")
```
How about 'Sleep for 45s for the attr_update_period to pass'

bhroam commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -132,8 +134,8 @@ def test_accrue_type(self):
+        self.server.expect(JOB, {"accrue_type": "1"}, id=jid4, max_attempts=1)
+        self.server.expect(JOB, "comment", op=SET, id=jid3, max_attempts=1)
+        self.server.expect(JOB, {"accrue_type": "1"}, id=jid3, max_attempts=1)
+        self.server.expect(JOB, {"accrue_type": "1"}, id=jid2, max_attempts=1)
```
If you are going to do max_attempts=1, then do a status(), and use asserts.  This means you'll do one qstat -f instead of 5.

agrawalravi90 replied:
status returns a dict, not a list, so I'll have to iterate over it, match the jid and then do asserts accordingly. I kind of feel like 5 qstats is not a big deal, we often do max_attempts=5 anyways for existence=False. So I'd like to keep it as is, this will make the test more readable, easier to modify in future. Let me know what you think.

agrawalravi90 commented:
@bhroam I've addressed your comments, removed sched_config's hidden opt_backfill_fuzzy_time, and enhanced qmgr sched object's opt_backfill_fuzzy to accept time in seconds. Please provide further feedback, thanks!

bhroam commented on src/server/svr_func.c:
```diff
@@ -7196,10 +7196,16 @@ int
+	strtol(str, &endp, 10);
+	if (*endp == '\0')
+		return PBSE_NONE;
+
```
This should be inside the if statement right below this.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openpbs/openpbs?pullRequest=1689) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/openpbs/openpbs?pullRequest=1689) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openpbs/openpbs?pullRequest=1689) it.</sub>

bhroam commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -516,6 +516,13 @@ def setUp(self):
+        for schedinfo in self.schedulers.values():
+            for schedname in schedinfo.keys():
+                self.server.manager(MGR_CMD_SET, SCHED, a, id=schedname)
+
```
Why are you turning off opt_backfill_fuzzy?  Plus, the reason I had to ask this means your comment is not helpful.  Either add why or remove it.

bhroam replied:
I am fine with this.  Make sure the opt_backfill_fuzzy test still passes.  It might have a test which tests the default behavior.

agrawalravi90 replied:
Thanks, I ran a regression after this and TestCalendaring passed. Let me know if you'd like me to run any other test as well.

bhroam replied:
Is that where the opt_backfill_fuzzy tests are?  I thought there was a test plan just for it.

agrawalravi90 replied:
that's all that i found. There were a couple in other test suites as well, but that was it.

agrawalravi90 commented:
> Why the need for the job_sort_formula global? The point of sc_attrs is to get rid of one-off globals like that.

It's because we have a server job_sort_formula, and a sched job_sort_formula, and one doesn't set/unset the other. So, if I set server's job_sort_formula and set sc_attrs' JSF to that value, then unset the server's JSF, sc_attrs' JSF doesn't get unset. So, it was either to create this global, or to change the behavior of PBS so that JSF of server sets/unsets JSF of all schedulers as well.

bhroam commented:
That's right, the JSF is special.  Man I hate it when things are special cases.  If JSF is not set on the scheduler, it uses the servers.  You can use the scheduler's to override the servers.  This means you need to know if it is set in the scheduler before you use the servers.  Ugh, you do need both.  Since we are going to be querying the server's JSF anyway, you could put the JSF back in the sinfo.  set it to scattrs by default.  If is not set when querying JSF on the server, use the server value.

bhroam commented on src/scheduler/server_info.c:
```diff
@@ -636,6 +636,12 @@ query_server_info(status *pol, struct batch_status *server)
+	if (sinfo->job_sort_formula == NULL && sc_attrs.job_sort_formula != NULL) {
+		sinfo->job_sort_formula = string_dup(sc_attrs.job_sort_formula);
+		if (sinfo->job_sort_formula == NULL)
+			return NULL;
```
If you return NULL, you'll leak the sinfo structure.  You need to free it first.

bhroam commented on src/scheduler/server_info.c:
```diff
@@ -2287,6 +2295,12 @@ dup_server_info(server_info *osinfo)
+	if (osinfo->job_sort_formula != NULL) {
+		nsinfo->job_sort_formula = string_dup(osinfo->job_sort_formula);
+		if (nsinfo->job_sort_formula == NULL)
+			return NULL;
```
Same here, you'll leak nsinfo.

nithinj commented on src/scheduler/fifo.h:
```diff
@@ -232,7 +232,7 @@ int scheduler_simulation_task(int pbs_sd, int debug);
 int update_svr_schedobj(int connector, int cmd, int alarm_time);
 
-int validate_sched_attrs(int connector);
+int set_validate_sched_attrs(int connector);
```
Use extern keyword

agrawalravi90 replied:
why?

nithinj replied:
It is part of our [coding guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1481670667/PBS+Pro+Coding+Standards). 

But it should be okay as none of the functions in this header file has extern in it.

nithinj commented on src/scheduler/resv_info.c:
```diff
@@ -1576,7 +1576,7 @@ confirm_reservation(status *policy, int pbs_sd, resource_resv *unconf_resv, serv
 		 */
 		snprintf(confirm_msg, LOG_BUF_SIZE, "%s:partition=%s", PBS_RESV_CONFIRM_SUCCESS,
-			 nsinfo->partition?nsinfo->partition:DEFAULT_PARTITION);
+			 sc_attrs.partition?sc_attrs.partition:DEFAULT_PARTITION);
```
Keep space around operator.

nithinj commented on test/tests/functional/pbs_sched_attr_updates.py:
```diff
@@ -0,0 +1,139 @@
+# Copyright (C) 1994-2020 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
```
copyright header has to change.

agrawalravi90 replied:
thanks for catching that!

agrawalravi90 commented:
@nithinj  thanks a lot for the review, I've addressed your comments, please provide further feedback.

agrawalravi90 commented:
Thanks guys, ill go ahead and rebase this + resolve conflicts

agrawalravi90 commented:
@bhroam  and @nithinj  I had to fix some job_run_wait related conflicts in scheduler, can you please look at send_runjob() and run_job() in fifo.c once and let me know if you are okay with the changes?

bhroam commented:
@agrawalravi90 I have two comments.  First, could you either rename run_job() runjob(), or rename send_runjob() send_run_job().  It just looks weird with one having an underscore and the other not.  My other comment is that you are now passing in sinfo into run_job() just for has_runjob_hook.  Why not just pass in has_runjob_hook.  If we need to pass more info in the future, we can turn the parameter into flags and has_runjob_hook into a bit.

nithinj commented on src/scheduler/fifo.c:
```diff
@@ -1288,22 +1294,22 @@ update_job_can_not_run(int pbs_sd, resource_resv *job, schd_error *err)
-	else if (rj_mode == RJ_RUNJOB_HOOK)
+	if (sc_attrs.runjob_mode == RJ_EXECJOB_HOOK)
+		return pbs_runjob(pbs_sd, jobid, execvnode, NULL);
+	else if (sc_attrs.runjob_mode == RJ_RUNJOB_HOOK && has_runjob_hook)
```
enclose the first check.

nithinj commented on src/scheduler/fifo.c:
```diff
@@ -1315,15 +1321,15 @@ send_runjob(int pbs_sd, enum runjob_mode rj_mode, char *jobid, char *execvnode)
  * @param[in]	rjob	-	the job to run
  * @param[in]	execvnode	-	the execvnode to run a multi-node job on
- * @param[in]	rj_mode	-	runjob mode to use
+ * @param[in]	has_runjob_hook	- does server have a runjob hook?
```
Indentation looks off.

nithinj commented on src/scheduler/fifo.c:
```diff
@@ -1616,10 +1622,10 @@ run_update_resresv(status *policy, int pbs_sd, server_info *sinfo,
 
 #ifdef NAS_CLUSTER /* localmod 125 */
-					ret = translate_runjob_return_code(pbsrc, resresv);
+					    ret = translate_runjob_return_code(pbsrc, resresv);
```
Indentation!

agrawalravi90 commented:
@nithinj thanks, i've addressed your comments, please let me know if it looks okay now.

agrawalravi90 commented:
@bhroam thanks! I think this is ready for merge then
